MuyGPyS/gp/distortion/anisotropic.py

```
@@ -0,0 +1,120 @@
+        self.length_scale = length_scales
+        if metric == "l2":
+            self._dist_fn = _l2
+        elif metric == "F2":
+            self._dist_fn = _F2
+        else:
+            raise ValueError(f"Metric {metric} is not supported!")
+
+    def __call__(self, diffs: mm.ndarray, **length_scales) -> mm.ndarray:
+        partial_name = "length_scale"
+        length_scale_array = [
+            value
+            for key, value in length_scales.items()
+            if key.startswith(partial_name)
+        ]
+        length_scale_array = mm.array(length_scale_array)
+        if (
+            diffs.shape[-1] != len(length_scale_array)
+            and len(length_scale_array) != 1
+        ):
+            raise ValueError(
+                f"Number of lengthscale parameters "
+                f"({len(length_scale_array)}) must match number of "
+                f"features ({diffs.shape[-1]}) or be 1 (Isotropic model)."
+            )
+        return self._dist_fn(diffs / length_scale_array)
+
+    def get_optim_params(
+        self,
+    ) -> Tuple[List[str], List[float], List[Tuple[float, float]]]:
+        """
+        Report lists of unfixed hyperparameter names, values, and bounds.
+
+        Returns
+        -------
+            names:
+                A list of unfixed hyperparameter names.
+            params:
+                A list of unfixed hyperparameter values.
+            bounds:
+                A list of unfixed hyperparameter bound tuples.
+        """
+        names: List[str] = []
+        params: List[float] = []
+        bounds: List[Tuple[float, float]] = []
+        partial_key = "length_scale"
+        for key in self.length_scale.keys():
+            if partial_key.lower() in key.lower():
+                append_optim_params_lists(
+                    self.length_scale[key],
+                    key,
+                    names,
+                    params,
+                    bounds,
+                )
```

This pattern is repeated frequently, which must be maintained and executed in multiple places. I think it would be better to enforce that `self.length_scale` is set to a `Dict[str,Hyperparameter]` with keys `length_scaleX` for `X` in `[0, ..., response_count]`. We should throw an error otherwise (and should have a test that makes sure that we do). This way we can be certain of what `self.length_scale` contains, which will make code here and elsewhere simpler.

---

To be honest I use regex as infrequently as possible, so I don't have a lot of guidance.